src/lib/features/ai/tools/search-flag.ts

```diff
@@ -0,0 +1,161 @@
+import type { RunnableToolFunctionWithParse } from 'openai/lib/RunnableFunction';
+import type { IUser } from '../../../types';
+import type { FeatureSearchService } from '../../feature-search/feature-search-service';
+import { normalizeQueryParams } from '../../feature-search/search-utils';
+
+type searchFlagOperationalParams = {
+    featureSearchService: FeatureSearchService;
+    user: IUser;
+};
+
+type searchFlagParams = {
+    query?: string;
+    project?: string;
+    type?: string;
+    tag?: string;
+    segment?: string;
+    createdAt?: string;
+    state?: string;
+    status?: string[];
+};
+
+type searchFlagFunctionParams = searchFlagOperationalParams & searchFlagParams;
+
+const searchFlagFunction = async ({
+    featureSearchService,
+    user,
+    query,
+    project,
+    type,
+    tag,
+    segment,
+    createdAt,
+    state,
+    status,
+}: searchFlagFunctionParams) => {
+    try {
+        const userId = user.id;
+        const {
+            normalizedQuery,
+            normalizedSortOrder,
+            normalizedOffset,
+            normalizedLimit,
+        } = normalizeQueryParams(
+            { query },
+            {
+                limitDefault: 50,
+                maxLimit: 100,
+            },
+        );
+        const normalizedStatus = status
+            ?.map((tag) => tag.split(':'))
+            .filter(
+                (tag) =>
+                    tag.length === 2 &&
+                    ['enabled', 'disabled'].includes(tag[1]),
+            );
+
+        const response = await featureSearchService.search({
+            searchParams: normalizedQuery,
+            project,
+            type,
+            userId,
+            tag,
+            segment,
+            state,
+            createdAt,
+            status: normalizedStatus,
+            offset: normalizedOffset,
+            limit: normalizedLimit,
+            sortOrder: normalizedSortOrder,
+            favoritesFirst: true,
+        });
+
+        return response;
+    } catch (error) {
+        return error;
+    }
+};
+
+export const searchFlag = ({
+    featureSearchService,
+    user,
+}: searchFlagOperationalParams): RunnableToolFunctionWithParse<searchFlagParams> => ({
+    type: 'function',
```

Comment:
This is pretty cool! How well does this work in practice? Does OpenAI use all of these terms effectively?

Reply:
In general, based on the hackathon, this worked pretty well, so it should! We can always fine tune if needed.

If we're specifically talking about this `search-flag` method I'm not sure because I haven't tried it yet 😄 It's a new tool to replace the old `get-flag` one. Good news is that we'll find out soon enough whether this works like we expect it to!